### ISSUE TITLE
Photoshop RemoteConnection's mask function is not working.

### ISSUE THREAD

**Issue by za-wa-n-go** [NimaNzrii/comfyui-photoshop#116]
I have selected an area in Photoshop, but the mask function is not working. When I create a selection in Photoshop and execute it on the ComfyUI side, two transparent layers are generated in Photoshop, but nothing is obtained on the ComfyUI side.

<img width="226" alt="image" src="https://github.com/NimaNzrii/comfyui-photoshop/assets/118154415/1d2824df-7334-43cb-90ef-edf2cacfd278">

**Comment by za-wa-n-go** [comment]
With the help of ChatGPT, it is now operational.
I would appreciate it if you could check the code shared below.
If there are no issues, I would be happy if you could adopt this code.

```
from PIL import Image, ImageOps
import numpy as np
import torch
import tempfile
import os
import base64
import subprocess

class PhotoshopConnections:
    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "Selection_To_Mask": ("BOOLEAN", {"default": False}),
                "password": ("STRING", {"default": "12341234"}),
                "Server": ("STRING", {"default": "127.0.0.1"}),
                "port": ("STRING", {"default": "49494"})
            }
        }

    RETURN_TYPES = ("IMAGE", "MASK")
    RETURN_NAMES = ("Photoshop Canvas", "Mask")
    FUNCTION = "PS_Execute"
    CATEGORY = "Photoshop"

    def PS_Execute(self, Selection_To_Mask, password, Server, port):
        try:
            from photoshop import PhotoshopConnection
        except ImportError:
            subprocess.run(["python", "-m", "pip", "uninstall", "photoshop"], check=True)
            subprocess.run(["python", "-m", "pip", "uninstall", "photoshop-connection"], check=True)
            subprocess.run(["python", "-m", "pip", "install", "photoshop-connection"], check=True)
        
        self.TmpDir = tempfile.gettempdir().replace("\\", "/")
        self.ImgDir = f"{self.TmpDir}/temp_image.jpg"
        self.MaskDir = f"{self.TmpDir}/temp_image_mask.jpg"
        
        ImgScript = f'''var saveFile = new File("{self.ImgDir}"); var jpegOptions = new JPEGSaveOptions(); jpegOptions.quality = 10; activeDocument.saveAs(saveFile, jpegOptions, true);'''
        
        Maskscript = f'''try{{var e=app.activeDocument,a=e.selection.bounds,t=e.activeHistoryState,i=e.artLayers.add(),s=new SolidColor,r=e.artLayers.add(),l=new SolidColor,c=new File("{self.MaskDir}"),n=new JPEGSaveOptions;function o(){{s.rgb.hexValue="000000",l.rgb.hexValue="FFFFFF",e.activeLayer=r,e.selection.fill(l),e.activeLayer=i,e.selection.selectAll(),e.selection.fill(s),n.quality=1,e.saveAs(c,n,!0),e.activeHistoryState=t}}o()}}catch(y){{File("{self.MaskDir}").remove()}}'''

        with PhotoshopConnection(password=password, host=Server, port=port) as ps_conn:
            ps_conn.execute(ImgScript)
            if Selection_To_Mask:
                ps_conn.execute(Maskscript)

        self.SendImg(Selection_To_Mask)
        return (self.image, self.mask, self.width, self.height)

    def SendImg(self, Selection_To_Mask):
        self.loadImg(self.ImgDir)
        self.image = self.i.convert('RGB')
        self.image = np.array(self.image).astype(np.float32) / 255.0
        self.image = torch.from_numpy(self.image)[None,]
        self.width, self.height = self.i.size
        
        if Selection_To_Mask:
            self.loadImg(self.MaskDir)
            self.i = ImageOps.exif_transpose(self.i)
            self.mask = np.array(self.i.getchannel('B')).astype(np.float32) / 255.0
            self.mask = torch.from_numpy(self.mask)
        else:
            self.i = Image.new(mode='RGB', size=(1, 1), color=(0, 0, 0))
            self.i = ImageOps.exif_transpose(self.i)
            self.mask = np.array(self.i.getchannel('B')).astype(np.float32) / 255.0
            self.mask = torch.from_numpy(self.mask)

    def loadImg(self, path):
        try:
            self.i = Image.open(path)
            self.i.verify()
            self.i = Image.open(path)
        except:
            self.i = Image.new(mode='RGB', size=(1, 1), color=(0, 0, 0))
        if not self.i:
            return

    @classmethod
    def IS_CHANGED(cls, ImgDir, MaskDir):
        with open(ImgDir, "rb") as img_file:
            if os.path.exists(MaskDir):
                with open(MaskDir, "rb") as mask_file:
                    return base64.b64encode(mask_file.read()).decode('utf-8') + base64.b64encode(img_file.read()).decode('utf-8')
            else:
                return base64.b64encode(img_file.read()).decode('utf-8')

NODE_CLASS_MAPPINGS = {
    "🔹 Photoshop RemoteConnection": PhotoshopConnections
}

NODE_DISPLAY_NAME_MAPPINGS = {
    "PhotoshopConnection": "🔹 Photoshop RemoteConnection"
}

```

**Comment by za-wa-n-go** [comment]
Six hours ago, an update made the mask function work, but a phenomenon is occurring where four empty layers are being generated.

<img width="233" alt="image" src="https://github.com/NimaNzrii/comfyui-photoshop/assets/118154415/f3b73a92-3d7f-45ad-a9d5-2ad1c762f34a">

**Comment by za-wa-n-go** [comment]
I tried to ensure that empty layers are no longer generated. Please check.

`from PIL import Image, ImageOps
import numpy as np
import torch
import tempfile
import os
import base64
import subprocess

class PhotoshopConnections:
    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "Selection_To_Mask": ("BOOLEAN", {"default": False}),
                "password": ("STRING", {"default": "12341234"}),
                "Server": ("STRING", {"default": "127.0.0.1"}),
                "port": ("STRING", {"default": "49494"})
            }
        }

    RETURN_TYPES = ("IMAGE", "MASK")
    RETURN_NAMES = ("Photoshop Canvas", "Mask")
    FUNCTION = "PS_Execute"
    CATEGORY = "Photoshop"

    def PS_Execute(self, Selection_To_Mask, password, Server, port):
        try:
            from photoshop import PhotoshopConnection
        except:
            subprocess.run(["python", "-m", "pip", "uninstall", "photoshop"], check=True)
            subprocess.run(["python", "-m", "pip", "uninstall", "photoshop-connection"], check=True)
            subprocess.run(["python", "-m", "pip", "install", "photoshop-connection"], check=True)
        
        self.TmpDir = tempfile.gettempdir().replace("\\", "/")
        self.ImgDir = f"{self.TmpDir}/temp_image.jpg"
        self.MaskDir = f"{self.TmpDir}/temp_image_mask.jpg"
        
        ImgScript = f'''var saveFile = new File("{self.ImgDir}"); var jpegOptions = new JPEGSaveOptions(); jpegOptions.quality = 10; activeDocument.saveAs(saveFile, jpegOptions, true);'''
        
        Maskscript = f'''
        function main() {{
            try {{
                var e = app.activeDocument;
                var a = e.selection.bounds;
                var t = e.activeHistoryState;
                var i = e.artLayers.add();
                var s = new SolidColor();
                var r = e.artLayers.add();
                var l = new SolidColor();
                var c = new File("{self.MaskDir}");
                var n = new JPEGSaveOptions();
                s.rgb.hexValue = "000000";
                l.rgb.hexValue = "FFFFFF";
                e.activeLayer = r;
                e.selection.fill(l);
                e.activeLayer = i;
                e.selection.selectAll();
                e.selection.fill(s);
                n.quality = 1;
                e.saveAs(c, n, true);
                e.activeHistoryState = t;
            }} catch (y) {{
                File("{self.MaskDir}").remove();
                console.log("Error in mask creation: " + y.toString());
            }}
        }}
        app.activeDocument.suspendHistory("Mask Applied", "main()");
        '''

        with PhotoshopConnection(password=password, host=Server, port=port) as ps_conn:
            ps_conn.execute(ImgScript)
            if Selection_To_Mask:
                ps_conn.execute(Maskscript)

        self.SendImg(Selection_To_Mask)
        return (self.image, self.mask, self.width, self.height)

    def SendImg(self, Selection_To_Mask):
        self.loadImg(self.ImgDir)
        self.image = self.i.convert('RGB')
        self.image = np.array(self.image).astype(np.float32) / 255.0
        self.image = torch.from_numpy(self.image)[None,]
        self.width, self.height = self.i.size
        
        if Selection_To_Mask:
            self.loadImg(self.MaskDir)
            self.i = ImageOps.exif_transpose(self.i)
            self.mask = np.array(self.i.getchannel('B')).astype(np.float32) / 255.0
            self.mask = torch.from_numpy(self.mask)
        else:
            self.i = Image.new(mode='RGB', size=(1, 1), color=(0, 0, 0))
            self.i = ImageOps.exif_transpose(self.i)
            self.mask = np.array(self.i.getchannel('B')).astype(np.float32) / 255.0
            self.mask = torch.from_numpy(self.mask)

    def loadImg(self, path):
        try:
            self.i = Image.open(path)
            self.i.verify()
            self.i = Image.open(path)
        except Exception as e:
            print("Error loading image: ", e)
            self.i = Image.new(mode='RGB', size=(1, 1), color=(0, 0, 0))

    @classmethod
    def IS_CHANGED(cls, ImgDir, MaskDir):
        with open(ImgDir, "rb") as img_file:
            if os.path.exists(MaskDir):
                with open(MaskDir, "rb") as mask_file:
                    return base64.b64encode(mask_file.read()).decode('utf-8') + base64.b64encode(img_file.read()).decode('utf-8')
            else:
                return base64.b64encode(img_file.read()).decode('utf-8')

NODE_CLASS_MAPPINGS = {
    "🔹 Photoshop RemoteConnection": PhotoshopConnections
}

NODE_DISPLAY_NAME_MAPPINGS = {
    "PhotoshopConnection": "🔹 Photoshop RemoteConnection"
}
`

**Comment by za-wa-n-go** [comment]
I tried to ensure that empty layers are no longer generated. Please check.

```
from PIL import Image, ImageOps
import numpy as np
import torch
import tempfile
import os
import base64
import subprocess

class PhotoshopConnections:
    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "Selection_To_Mask": ("BOOLEAN", {"default": False}),
                "password": ("STRING", {"default": "12341234"}),
                "Server": ("STRING", {"default": "127.0.0.1"}),
                "port": ("STRING", {"default": "49494"})
            }
        }

    RETURN_TYPES = ("IMAGE", "MASK")
    RETURN_NAMES = ("Photoshop Canvas", "Mask")
    FUNCTION = "PS_Execute"
    CATEGORY = "Photoshop"

    def PS_Execute(self, Selection_To_Mask, password, Server, port):
        try:
            from photoshop import PhotoshopConnection
        except:
            subprocess.run(["python", "-m", "pip", "uninstall", "photoshop"], check=True)
            subprocess.run(["python", "-m", "pip", "uninstall", "photoshop-connection"], check=True)
            subprocess.run(["python", "-m", "pip", "install", "photoshop-connection"], check=True)
        
        self.TmpDir = tempfile.gettempdir().replace("\\", "/")
        self.ImgDir = f"{self.TmpDir}/temp_image.jpg"
        self.MaskDir = f"{self.TmpDir}/temp_image_mask.jpg"
        
        ImgScript = f'''var saveFile = new File("{self.ImgDir}"); var jpegOptions = new JPEGSaveOptions(); jpegOptions.quality = 10; activeDocument.saveAs(saveFile, jpegOptions, true);'''
        
        Maskscript = f'''
        function main() {{
            try {{
                var e = app.activeDocument;
                var a = e.selection.bounds;
                var t = e.activeHistoryState;
                var i = e.artLayers.add();
                var s = new SolidColor();
                var r = e.artLayers.add();
                var l = new SolidColor();
                var c = new File("{self.MaskDir}");
                var n = new JPEGSaveOptions();
                s.rgb.hexValue = "000000";
                l.rgb.hexValue = "FFFFFF";
                e.activeLayer = r;
                e.selection.fill(l);
                e.activeLayer = i;
                e.selection.selectAll();
                e.selection.fill(s);
                n.quality = 1;
                e.saveAs(c, n, true);
                e.activeHistoryState = t;
            }} catch (y) {{
                File("{self.MaskDir}").remove();
                console.log("Error in mask creation: " + y.toString());
            }}
        }}
        app.activeDocument.suspendHistory("Mask Applied", "main()");
        '''

        with PhotoshopConnection(password=password, host=Server, port=port) as ps_conn:
            ps_conn.execute(ImgScript)
            if Selection_To_Mask:
                ps_conn.execute(Maskscript)

        self.SendImg(Selection_To_Mask)
        return (self.image, self.mask, self.width, self.height)

    def SendImg(self, Selection_To_Mask):
        self.loadImg(self.ImgDir)
        self.image = self.i.convert('RGB')
        self.image = np.array(self.image).astype(np.float32) / 255.0
        self.image = torch.from_numpy(self.image)[None,]
        self.width, self.height = self.i.size
        
        if Selection_To_Mask:
            self.loadImg(self.MaskDir)
            self.i = ImageOps.exif_transpose(self.i)
            self.mask = np.array(self.i.getchannel('B')).astype(np.float32) / 255.0
            self.mask = torch.from_numpy(self.mask)
        else:
            self.i = Image.new(mode='RGB', size=(1, 1), color=(0, 0, 0))
            self.i = ImageOps.exif_transpose(self.i)
            self.mask = np.array(self.i.getchannel('B')).astype(np.float32) / 255.0
            self.mask = torch.from_numpy(self.mask)

    def loadImg(self, path):
        try:
            self.i = Image.open(path)
            self.i.verify()
            self.i = Image.open(path)
        except Exception as e:
            print("Error loading image: ", e)
            self.i = Image.new(mode='RGB', size=(1, 1), color=(0, 0, 0))

    @classmethod
    def IS_CHANGED(cls, ImgDir, MaskDir):
        with open(ImgDir, "rb") as img_file:
            if os.path.exists(MaskDir):
                with open(MaskDir, "rb") as mask_file:
                    return base64.b64encode(mask_file.read()).decode('utf-8') + base64.b64encode(img_file.read()).decode('utf-8')
            else:
                return base64.b64encode(img_file.read()).decode('utf-8')

NODE_CLASS_MAPPINGS = {
    "🔹 Photoshop RemoteConnection": PhotoshopConnections
}

NODE_DISPLAY_NAME_MAPPINGS = {
    "PhotoshopConnection": "🔹 Photoshop RemoteConnection"
}

```

**Comment by Anti-Liang** [comment]
maybe you have to try it on ENGver photoshop...

**Comment by za-wa-n-go** [comment]
This is probably not the main node for this custom node, so I decided to prepare it myself based on this. I will close it!
